### PR TITLE
use nginx as HTTP proxy

### PR DESCRIPTION
**Pull request by sqs**
Proposal: Remove anything from site config that could be done fairly easily by nginx. I think this includes things like:

- httpToHttpsRedirect
- httpStrictTransportSecurity
- tlsCert
- tlsKey
- tls.letsencrypt
- canonicalURLRedirect

Reason: Many of our big deployments use nginx anyway, so we *should* consider it as a first-class thing. This lets us remove code (that is nuanced/has large bug surface area). Also it's easier to Google "nginx how to xyz" than "sourcegraph how to xyz" where xyz is some http/https/etc.-related thing.

Contingent on: Finding a way to replace all those configs above with a simple nginx config snippet; and making this work cleanly in both sourcegraph/server and cluster.

> This PR updates the CHANGELOG.md file to describe any user-facing changes.

TODOs:

- [x] Update Server procfile to run nginx and document how customers can change it
- [x] Add sentence to docs that says for cluster deployments use nginx ingress
- [ ] Mention nginx in deploy-sourcegraph docs (but can defer full documentation until after 3.0) #1609 
- [ ] Ensure this works when deployed to dogfood - blocked on #1610
- [ ] Ensure this works when deployed to Sourcegraph.com - blocked on #1610
- [ ] Write test plan

—-

Issues that would be much easier if we used nginx (incomplete list):

- https://github.com/sourcegraph/sourcegraph/issues/969#issuecomment-438625634
- https://github.com/sourcegraph/sourcegraph/issues/807

### CODE REVIEW

**Comment by keegancsmith** [comment]
Add me back as a reviewer when its ready.

**Comment by sqs** [comment]
@keegancsmith 	This is ready for review.

File: dev/launch.sh [hunk 1]
```diff
@@ -69,9 +69,10 @@ export DEPLOY_TYPE=dev
 export ZOEKT_HOST=localhost:6070
 
 # webpack-dev-server is a proxy running on port 3080 that (1) serves assets, waiting to respond
-# until they are (re)built and (2) otherwise passes through to Sourcegraph running on port
-# 3081. That is why Sourcegraph listens on 3081 despite the externalURL having port 3080.
-export SRC_HTTP_ADDR=":3081"
+# until they are (re)built and (2) otherwise proxies to nginx running on port 3081 (which proxies to
+# Sourcegraph running on port 3082). That is why Sourcegraph listens on 3081 despite the externalURL
```
Comment (nicksnyder) [inline dev/launch.sh]:
```suggestion
# Sourcegraph running on port 3082). That is why Sourcegraph listens on 3082 despite the externalURL
```

**Comment by keegancsmith** [comment]
FYI I have this on my TODO list to review tomorrow.

**Comment by nicksnyder** [comment]
@keegancsmith is going to own testing this and getting it merged by EOW

**Comment by nicksnyder** [comment]
Any update here?

**Comment by keegancsmith** [comment]
Yes, will merge today. Still missing server image integration, otherwise shipping this may break dogfood.

**Comment by keegancsmith** [comment]
Documentation isn't ready, will do another PR. But for now this is ready for dogfood so wanna test that.

**Pull request merged**